lib/multicell_location/services/nrf_cloud_integration.c

```diff
@@ -60,6 +69,48 @@ const char *location_service_get_certificate_nrf_cloud(void)
 	return tls_certificate;
 }
 
+#if defined(CONFIG_NRF_CLOUD_MQTT)
+static void location_service_location_ready_cb(const struct nrf_cloud_cell_pos_result *const result)
+{
+	if (result != NULL) {
+		nrf_cloud_location.latitude = result->lat;
+		nrf_cloud_location.longitude = result->lon;
+		nrf_cloud_location.accuracy = (double)result->unc;
+	}
+	k_sem_give(&location_ready);
+}
+
+int location_service_get_cell_location_nrf_cloud(
+	const struct lte_lc_cells_info *cell_data,
+	char * const rcv_buf,
+	const size_t rcv_buf_len,
+	struct multicell_location *const location)
+{
+	ARG_UNUSED(rcv_buf);
+	ARG_UNUSED(rcv_buf_len);
+
+	int err;
+
+	err = nrf_cloud_cell_pos_request(cell_data, true, location_service_location_ready_cb);
+	if (err) {
+		LOG_ERR("Failed to request cell pos data, error: %d", err);
+		return err;
+	}
+
+	LOG_INF("Cellular positioning request sent");
+
+	if (k_sem_take(&location_ready, K_SECONDS(20)) == -EAGAIN) {
```

Comment:
This is a bit problematic when multicell lib is used from location lib. It does not obey the location lib timeout. If I put K_FOREVER here it will get stuck forever here in case the MQTT connection has dropped (which happens soon with some roaming SIM cards). How to solve it elegantly?

Reply:
How did this work with REST? We do have something similar with REST but it's just hidden inside Zephyr's http_client, i.e., http_client will wait for the data to arrive.
@jhirsi can probably shed some light on how did the timeout work with that after he moved it to different work queue than the rest of location library operations.

But, even if location lib would work somehow, multicell_location_get() still has to return at some point as there is no cancel function.

Reply:
Yeah, this is tricky. There are different levels of timeouts.
Location api is having a timeout and timeout worker is running in different thread where semaphores are waiting.
http_client is having own timeout and waits for the data within http_client_req() call.
Both are blocking/synch apis.
If location api timeout happens when waiting for data in http_client,
then operation is just cancelled but the data waiting expires only
when http_client timeout is triggered. And because there isn't timeout in multicell api (?),
location api timeout can happen 1st and the data might still come at background after the location api timeout but it is thrown away.
timeout in multicell api might improve so that location api could pass the remaining time (from location api timer after ncellmeas) to multicell lib
and that time could be used for waiting here as well.

If not wanting to change the api at this point, maybe at minimum we should have a Kconfig for the timeout (assuming that there isn't in MQTT side? Then that would be analogue with CONFIG_REST_CLIENT_REQUEST_TIMEOUT.

Reply:
I understood we will try to introduce timeout parameter into the interface of multicell_location not in NCS 1.8.0 but in the following release, so we can leave it as it is now....unless anybody wants to change the actual value?

Reply:
Agreed to leave it at 20 seconds